### ISSUE TITLE
Large regression in word boundary regexes

### ISSUE THREAD

**Issue by rctcwyvrn** [apple/swift-experimental-string-processing#567]
I was running the benchmarker for this regex `#"<(\w*)\b[^>]*>(.*?)<\/\1>"#` which uses `\b` to match the end of a html tag and noticed it was running _really_ slow

ed842cbb844a670314f22b37936b654748fad91d
```
Running
- htmlAll 11.8ms
```
`main`
```
Running
- htmlAll 3.08s
```

Some amount of regression was expected with the implementation of the new word breaking algorithm but a 300x slowdown seems unacceptable. A quick profile shows that ~90% of the time is spent in `String._wordIndex(after:)`

cc @Azoy @milseman

**Comment by milseman** [comment]
@Azoy is this because the SPI is inefficient, or any thoughts on what to do here?

**Comment by Azoy** [comment]
yeah the current implementation of `String.isOnWordBoundary` in this repo is really inefficient and was fully expecting perf to be pretty bad. Once `_nearestWordIndex(atOrBelow:)` is fixed, I think this operation will get considerably faster.
